p4-16/spec/P4-16-spec.mdk

```diff
@@ -7182,28 +7181,44 @@ The evaluation of a P4 program is done in two stages:
   executed to completion, in isolation, when it receives control from the
   architecture
 
-## Compile-time known values { #sec-ct-constants }
+## Compile-time known and local compile-time known values { #sec-ct-constants }
 
-The following are compile-time known values:
+Certain expressions in a P4 program have the property that their value
+can be determined at compile time. Moreover, for some of these
+expressions, their value can be determined only using information in
+the current scope. We call these _compile-time known values_ and
+_local compile-time known values_ respectively.
+
+The following are local compile-time known values:
 
 - Integer literals, Boolean literals, and string literals.
-- Identifiers declared in an `error`, `enum`, or `match_kind`
-  declaration.
+- Identifiers declared in an `error`, `enum`, or `match_kind` declaration.
 - The `default` identifier.
 - The `size` field of a value with type header stack.
 - The `_` identifier when used as a `select` expression label
-- Identifiers that represent declared types, actions, tables, parsers,
-  controls, or packages.
-- Tuple expression where all components are compile-time known values.
-- Structure-valued expressions, where all fields are compile-time
-  known values.
-- Instances constructed by instance declarations (Section
-  [#sec-instantiations]) and constructor invocations.
-- The following expressions (`+`, `-`, `*`, `/ `, `%`, `cast`, `!`, `&`, `|`, `&&`, `||`, `<< `, `>> `, `~` `, `\ `>`, `<`, `==`, `!=`, `<=`, `>=`, `++`, `[:]`, `?:`)
-  when their operands are all compile-time known values.
+- Identifiers that represent declared types, actions, tables, parsers, controls, or packages.
+- Tuple expression where all components are local compile-time known values.
+- Structure-valued expressions, where all fields are local compile-time known values.
+- Instances constructed by instance declarations (Section [#sec-instantiations]) and constructor invocations.
+- The following expressions (`+`, `-`, `*`, `/ `, `%`, `cast`, `!`, `&`, `|`, `&&`, `||`, `<< `, `>> `, `~` `, `\ `>`, `<`, `==`, `!=`, `<=`, `>=`, `++`, `[:]`, `?:`) when their operands are all local compile-time known values.
 - Identifiers declared as constants using the `const` keyword.
-- Expressions of the form `e.minSizeInBits()`, `e.minSizeInBytes()`,
-  `e.maxSizeInBits()` and `e.maxSizeInBytes()`
+- Expressions of the form `e.minSizeInBits()`, `e.minSizeInBytes()`, `e.maxSizeInBits()` and `e.maxSizeInBytes()`
+
+The following are compile-time known values:
+
+- All local compile-time known values.
+- Tuple expression where all components are compile-time known values.
+- Structure-valued expressions, where all fields are compile-time known values.
+- The following expressions (`+`, `-`, `*`, `/ `, `%`, `cast`, `!`, `&`, `|`, `&&`, `||`, `<< `, `>> `, `~` `, `\ `>`, `<`, `==`, `!=`, `<=`, `>=`, `++`, `[:]`, `?:`) when their operands are all compile-time known values.
```

Comment:
`^` was missing from the list before but I suspect it should be added.
Also this seems to be a bit off (which it was before too): ```
`~` `, `\ `>`, 
```
maybe it should have been:
```
`~`, `/`, `>`, 
```
The other binary expressions that are missing are: `|+|` and `|-|` (sat add and sat sub). They were missing before too.

Reply:
ok, the formating inside the block still messed up too.

Reply:
Well spotted. I think I've fixed these now. Thanks!